msm/src/serialization/constraints.rs

```diff
@@ -1,14 +1,88 @@
-use ark_ff::PrimeField;
+use ark_ff::{Field, PrimeField, Zero};
 use kimchi::circuits::{
-    expr::{ConstantExpr, ConstantTerm, Expr, ExprInner, Variable},
+    expr::{ChallengeTerm, ConstantExpr, ConstantTerm, Expr, ExprInner, Variable},
     gate::CurrOrNext,
 };
 
-use crate::{columns::Column, expr::E, serialization::N_INTERMEDIATE_LIMBS, N_LIMBS};
+use crate::{
+    columns::Column,
+    expr::{curr_cell, E},
+    serialization::N_INTERMEDIATE_LIMBS,
+    MVLookupTableID as _, N_LIMBS,
+};
 
 use super::Lookup;
 use super::{interpreter::InterpreterEnv, LookupTable};
 
+/// Compute the following constraint:
+/// ```text
+///                     lhs
+///    |------------------------------------------|
+///    |                           denominators   |
+///    |                         /--------------\ |
+/// column * (\prod_{i = 1}^{N} (\beta + f_{i}(X))) =
+/// \sum_{i = 1}^{N} m_{i} * \prod_{j = 1, j \neq i}^{N} (\beta + f_{j}(X))
+///    |             |--------------------------------------------------|
+///    |                             Inner part of rhs                  |
+///    |                                                                |
+///    |                                                               /
+///     \                                                             /
+///      \                                                           /
+///       \---------------------------------------------------------/
+///                           rhs
+/// ```
+pub fn combine_lookups<F: Field>(column: Column, lookups: Vec<Lookup<E<F>>>) -> E<F> {
+    let joint_combiner = {
+        let joint_combiner = ConstantExpr::from(ChallengeTerm::JointCombiner);
+        E::Atom(ExprInner::Constant(joint_combiner))
+    };
+    let beta = {
+        let beta = ConstantExpr::from(ChallengeTerm::Beta);
+        E::Atom(ExprInner::Constant(beta))
+    };
+
+    // Compute (\beta + f_{i}(X)) for each i.
+    // Note that f_i(X) = x_{0} + r x_{1} + ... r^{N} x_{N} + r^{N + 1} table_id
+    let denominators = lookups
+        .iter()
+        .map(|x| {
+            let combined_value = (x
+                .value
+                .iter()
+                .rev()
+                .fold(E::zero(), |acc, y| acc * joint_combiner.clone() + y.clone())
+                * joint_combiner.clone())
+                + x.table_id.to_constraint();
+            beta.clone() + combined_value
+        })
+        .collect::<Vec<_>>();
+    // Compute `column * (\prod_{i = 1}^{N} (\beta + f_{i}(X)))`
+    let lhs = denominators
+        .iter()
+        .fold(curr_cell(column), |acc, x| acc * x.clone());
+    let rhs = lookups
+        .into_iter()
+        .enumerate()
+        .map(|(i, x)| {
+            denominators.iter().enumerate().fold(
+                // Compute individual \sum_{j = 1, j \neq i}^{N} f_{j}(X)
+                // This is the inner part of rhs. It multiplies with m_{i}
```

Comment:
Perhaps it's not clear to me because it is not very imperative-oriented, but can you clarify which of these ones is the `m_{i}`'s and which are the `(beta + f_j)`?

Reply:
`numerator` is the `m_{i}`.
`beta + f_{j}` is from the denominator.